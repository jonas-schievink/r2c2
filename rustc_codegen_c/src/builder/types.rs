//! C type definitions and combinators.
//!
//! Make sure all C types have to be created through the `Factory`.

use utils::{StringWriter, WriteStr};

use std::fmt;
use std::io::{self, Write};

/// A function signature.
#[derive(Copy, Clone)]
pub struct FnSig<'a> {
    pub ret: TypeRef<'a>,
    pub args: &'a [TypeRef<'a>],
}

impl FnSig<'_> {
    /// (Forward-)Declares a function with this signature (partially).
    ///
    /// The output will not contain a trailing `;`. This makes this function
    /// also useful for declaring function pointer variables/static and casts.
    pub fn declare(&self, name: impl Into<String>, w: &mut impl WriteStr) -> io::Result<()> {
        // this is perverse
        let mut buf = StringWriter(name.into());
        buf.push('(');
        for (i, arg) in self.args.iter().enumerate() {
            if i != 0 {
                buf.push_str(", ");
            }

            arg.declare_variable(format!("_{}", i + 1), &mut buf)?;
        }
        if self.args.is_empty() {
            // `void` ensures the compiler doesn't let us call the function with arguments
            buf.push_str("void");
        }
        buf.push(')');

        self.ret.declare_variable(&buf, w)
    }
}

#[derive(Copy, Clone)]
pub enum Type<'a> {
    Bool,
    U8,
    I8,
    U16,
    I16,
    U32,
    I32,
    U64,
    I64,
    Float,
    Double,
    Pointer(&'a Type<'a>),
    /// `intptr_t` aka `isize`.
    IntPtr,
    /// `uintptr_t` aka `usize`
    UintPtr,
    Array {
        ty: &'a Type<'a>,
        len: usize,
    },
    FunctionPointer(FnSig<'a>),
    Struct {
        name: &'a str,
    },

    // Incomplete types
    Void,
    /// A forward-declared struct with unspecified fields.
    FwdStruct {
        /// The struct's name (mangled Rust name).
        name: &'a str,
    },
    FwdUnion {
        name: &'a str,
    },
}

impl Type<'_> {
    fn declare_variable<W: WriteStr>(&self, name: &str, w: &mut W) -> io::Result<()> {
        let simplety = match self {
            Type::Bool => "bool",
            Type::U8 => "uint8_t",
            Type::I8 => "int8_t",
            Type::U16 => "uint16_t",
            Type::I16 => "int16_t",
            Type::U32 => "uint32_t",
            Type::I32 => "int32_t",
            Type::U64 => "uint64_t",
            Type::I64 => "int64_t",
            Type::Float => "float",
            Type::Double => "double",
            Type::IntPtr => "intptr_t",
            Type::UintPtr => "uintptr_t",
            Type::Pointer(pointee) => return pointee.declare_variable(&format!("* {}", name), w),
            Type::Array { ty, len } => {
                return ty.declare_variable(&format!("{}[{}]", name, len), w)
            }
            Type::FunctionPointer(sig) => {
                // fn pointer syntax is like fn declaration syntax except the variable name is put
                // in `(*NAME)` where the fn name is
                return sig.declare(format!("(*{})", name), w);
            }
            Type::FwdStruct { name: ty } | Type::Struct { name: ty } => {
                return write!(w, "struct {} {}", ty, name)
            }
            Type::FwdUnion { name: ty } => return write!(w, "union {} {}", ty, name),
            // the void case is needed to declare functions and fn pointers
            Type::Void => return write!(w, "void {}", name),
        };

        write!(w, "{} {}", simplety, name)
    }
}

impl fmt::Debug for Type<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut buf = StringWriter(String::new());
        self.declare_variable("", &mut buf).unwrap();
        f.write_str(&buf)
    }
}

/// Trait abstracting over `TypeRef` and `IncompleteTypeRef`, used by methods
/// that can work with either.
pub trait AsType<'a> {
    fn as_type(&self) -> &'a Type<'a>;
}

/// Reference to a complete C type.
#[derive(Debug, Copy, Clone)]
pub struct TypeRef<'a>(pub &'a Type<'a>);

impl<'a> AsType<'a> for TypeRef<'a> {
    fn as_type(&self) -> &'a Type<'a> {
        self.0
    }
}

impl TypeRef<'_> {
    /// Declares a variable of this type and prints the declaration to `f`.
    ///
    /// Output is normally of the form `type name`. Note that no trailing `;` is
    /// printed.
    ///
    /// This can be used to declare local and global variables as well as struct
    /// fields.
    pub fn declare_variable<W: WriteStr>(
        &self,
        name: impl AsRef<str>,
        w: &mut W,
    ) -> io::Result<()> {
        self.0.declare_variable(name.as_ref(), w)
    }
}

/// Reference to an incomplete C type.
#[derive(Debug, Copy, Clone)]
pub struct IncompleteTypeRef<'a>(pub &'a Type<'a>);

impl<'a> AsType<'a> for IncompleteTypeRef<'a> {
    fn as_type(&self) -> &'a Type<'a> {
        self.0
    }
}
