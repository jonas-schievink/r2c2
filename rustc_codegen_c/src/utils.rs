use std::io::{self, Write};
use std::fmt::{self, Display};
use std::str;
use std::ops::Deref;
use std::ops::DerefMut;
use rustc::ty::TyCtxt;

// Import the host/target triple, set by the build script
include!(concat!(env!("OUT_DIR"), "/triples.rs"));

/// Extension methods for `Result<T, E>`.
pub trait ResultExt<T, E> {
    /// Turns the error into a `String`.
    fn err_to_string(self) -> Result<T, String>
    where
        E: ToString;
}

impl<T, E: Display> ResultExt<T, E> for Result<T, E> {
    fn err_to_string(self) -> Result<T, String> {
        self.map_err(|e| e.to_string())
    }
}

/// Obtains the version information of the system's C compiler.
///
/// TODO: Upstream this to the `cc` crate
pub fn cc_version() -> Result<String, String> {
    // FIXME: cc::Error doesn't impl Display and Error
    let mut compiler = cc::Build::new()
        .try_get_compiler()
        .map_err(|e| format!("{:?}", e))?
        .to_command();
    let output = compiler.arg("--version").output().err_to_string()?;

    let mut out = String::from_utf8(output.stdout).err_to_string()?;
    out.push_str(str::from_utf8(&output.stderr).err_to_string()?);

    if output.status.success() {
        Ok(out)
    } else {
        Err(out)
    }
}


/// Trait for writing UTF-8 data to a sink.
pub trait WriteStr {
    fn write_fmt(&mut self, args: fmt::Arguments) -> io::Result<()>;

    fn write_str(&mut self, s: &str) -> io::Result<()> {
        write!(self, "{}", s)
    }
}

impl<W: Write> WriteStr for W {
    fn write_fmt(&mut self, args: fmt::Arguments) -> io::Result<()> {
        <W as Write>::write_fmt(self, args)
    }
}

pub struct StringWriter(pub String);

impl WriteStr for StringWriter {
    fn write_fmt(&mut self, args: fmt::Arguments) -> io::Result<()> {
        Ok(<String as fmt::Write>::write_fmt(&mut self.0, args).expect("writing to string failed"))
    }
}

impl Deref for StringWriter {
    type Target = String;

    fn deref(&self) -> &String {
        &self.0
    }
}

impl DerefMut for StringWriter {
    fn deref_mut(&mut self) -> &mut String {
        &mut self.0
    }
}

impl AsRef<str> for StringWriter {
    fn as_ref(&self) -> &str {
        &self.0
    }
}
