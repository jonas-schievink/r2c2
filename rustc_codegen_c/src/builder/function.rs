//! Function builder.

use super::types::{FnSig, TypeRef};
use super::Symbol;
use utils::WriteStr;

use hashbrown::HashSet;
use toolshed::Arena;

use std::io;

/// A local variable.
///
/// These can be created in 2 ways: By declaring a function taking arguments
/// (which creates `Variable`s corresponding to the arguments), or by calling
/// `declare_variable` on an existing `FunctionBuilder`, which declares a new
/// local variable inside the function.
#[derive(Debug, Copy, Clone)]
pub struct Variable<'a> {
    name: &'a str,
    ty: TypeRef<'a>,
}

/// Builder for function bodies.
///
/// Created by `TranslationUnitBuilder::define_function`.
pub struct FunctionBuilder<'a, W: WriteStr> {
    /// Output statements are written to this writer.
    writer: &'a mut W,
    arena: &'a Arena,
    declared_locals: HashSet<String>,
    finished: bool,

    /// Function arguments, available as declared `Variable`s.
    pub args: &'a [Variable<'a>],
}

impl<'a, W: WriteStr> FunctionBuilder<'a, W> {
    /// Creates a function builder, writing the function's header to `writer`.
    ///
    /// **Note**: You *must* call `finish` on the builder to close the function
    /// being built, or this will panic on drop.
    pub fn create(
        writer: &'a mut W,
        arena: &'a Arena,
        name: Symbol<'_>,
        proto: FnSig<'a>,
    ) -> io::Result<Self> {
        proto.declare(name.mangled, writer)?;
        writeln!(writer)?;
        writeln!(writer, "{{")?;
        Ok(Self {
            writer,
            arena,
            declared_locals: HashSet::new(),
            finished: false,
            args: arena.alloc_slice(
                &proto
                    .args
                    .iter()
                    .enumerate()
                    .map(|(i, &ty)| Variable {
                        name: arena.alloc_str(&format!("_{}", i)),
                        ty,
                    })
                    .collect::<Vec<_>>(),
            ),
        })
    }

    /// Closes the in-progress function definition, consuming `self`.
    pub fn finish(mut self) -> io::Result<()> {
        writeln!(self.writer, "}}")?;
        writeln!(self.writer)?;
        self.finished = true;
        Ok(())
    }

    fn indent(&mut self) -> io::Result<()> {
        write!(self.writer, "    ")
    }

    /// Declares a new local variable at the current position.
    pub fn declare_variable<'b>(
        &mut self,
        name: impl AsRef<str>,
        ty: TypeRef<'a>,
        comment: impl Into<Option<&'b str>>,
    ) -> io::Result<Variable<'a>> {
        let name = name.as_ref();
        self.indent()?;
        ty.declare_variable(name, self.writer)?;
        write!(self.writer, ";")?;
        if let Some(comment) = comment.into() {
            write!(self.writer, "  /* {} */", comment)?;
        }
        writeln!(self.writer)?;

        Ok(Variable {
            name: self.arena.alloc_str(name),
            ty,
        })
    }
}

impl<'a, W: WriteStr> Drop for FunctionBuilder<'a, W> {
    fn drop(&mut self) {
        if !self.finished {
            panic!("FunctionBuilder dropped without calling `finish`");
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use builder::test::compile_test;

    #[test]
    fn declare_locals() {
        compile_test("declare_locals", |tu| {
            let a = Arena::new();
            let sig = tu.fn_sig(None, &[]);
            let double = tu.double();
            let fnptr = tu.fn_ptr(sig);
            let mut f = tu.define_function(&a, Symbol::test("declare_locals"), sig)?;
            f.declare_variable("dbl", double, None)?;
            f.declare_variable("f", fnptr, Some("i'm a function pointer with a comment"))?;
            f.finish()?;
            Ok(())
        });
    }
}
