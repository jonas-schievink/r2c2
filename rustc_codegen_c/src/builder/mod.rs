//! A higher-level builder interface to C code generation.
//!
//! Notes on C namespaces and item naming:
//! * structs and unions share a namespace
//! * Local variables shadow functions
//! * Rust and C use the same identifier syntax, so escaping is difficult
//! * Keep all global identifiers as-is (should work due to mangling) and rename
//!   colliding locals

pub mod function;
pub mod test;
pub mod types;

use self::function::FunctionBuilder;
use self::types::{AsType, FnSig, IncompleteTypeRef, Type, TypeRef};
use utils::WriteStr;

use hashbrown::hash_map::{HashMap, RawEntryMut};
use hashbrown::HashSet;
use toolshed::Arena;

use bitflags::_core::fmt::{Error, Formatter};
use rustc::ty::{Instance, SymbolName, TyCtxt};
use rustc_index::{newtype_index, vec::IndexVec};
use std::borrow::Cow;
use std::fmt;
use std::io::{self, Write};
use std::rc::Rc;
use syntax::symbol::{Symbol, SymbolStr};

/// A Rust/C function or type name.
pub struct Name<'a> {
    rust: Cow<'a, str>,
    mangled: Symbol,
}

impl<'a> Name<'a> {
    /// Create a `Symbol` for a monomorphized `Instance`.
    pub fn for_instance<'tcx>(inst: &Instance<'tcx>, tcx: TyCtxt<'tcx>) -> Self {
        Self {
            rust: inst.to_string().into(),
            mangled: tcx.symbol_name(*inst).name,
        }
    }

    /// A test symbol that doesn't have an associated Rust name (used only by unit tests).
    pub fn test(c_name: &'a str) -> Self {
        Self {
            rust: "(test-symbol: no associated Rust name)".into(),
            mangled: Symbol::intern(c_name),
        }
    }

    /// Returns the Rust path to the symbol (possibly containing a comment).
    pub fn rust(&self) -> &str {
        &self.rust
    }

    /// Returns the mangled C name of this symbol.
    pub fn mangled(&self) -> SymbolStr {
        self.mangled.as_str()
    }
}

/// A builder for C translation units.
///
/// A "translation unit" is a single `.c` file and corresponds directly to a
/// codegen unit in Rust.
pub struct TranslationUnitBuilder<'a, W: WriteStr> {
    /// Output file writer.
    writer: W,

    /// Arena storing TU-wide data (types, type names, etc.).
    arena: &'a Arena,

    /// `struct`s and `union`s that have been defined (not declared) so far.
    defd_types: HashSet<&'a str>,
}

impl<'a, W: WriteStr> TranslationUnitBuilder<'a, W> {
    /// Creates a new C file builder that writes its output to `writer`.
    ///
    /// This will also write the preamble to the writer, which includes standard
    /// files that are assumed to be present by the rest of the code generator.
    pub fn create(arena: &'a Arena, writer: W) -> io::Result<Self> {
        let mut this = Self {
            writer,
            arena,
            defd_types: HashSet::new(),
        };

        // Include a few standard headers the code generator relies on
        this.comment(&format!(
            "generated by r2c2 version {}",
            env!("CARGO_PKG_VERSION")
        ))?;
        this.include("stdint.h")?;
        this.include("stdbool.h")?;
        this.separator("END PREAMBLE")?;

        Ok(this)
    }

    /// Returns the output file writer associated with this TU builder.
    pub fn writer(&mut self) -> &mut W {
        &mut self.writer
    }

    pub fn into_writer(self) -> W {
        self.writer
    }

    fn include(&mut self, sys: &str) -> io::Result<()> {
        writeln!(self.writer, "#include <{}>", sys)
    }

    fn comment(&mut self, comment: &str) -> io::Result<()> {
        for line in comment.lines() {
            writeln!(self.writer, "/* {} */", line.replace("*/", "*-/"))?;
        }
        Ok(())
    }

    /// Emits 3 comment lines to separate different pieces of output.
    pub fn separator(&mut self, name: &str) -> io::Result<()> {
        writeln!(self.writer)?;
        self.comment(&" ".repeat(50))?;

        let spaces = 50usize.saturating_sub(name.len());
        let text = format!(
            "{}{}{}",
            " ".repeat(spaces / 2),
            name,
            " ".repeat((spaces + 1) / 2)
        );
        self.comment(&text)?;

        self.comment(&" ".repeat(50))?;
        writeln!(self.writer)?;
        Ok(())
    }
}

macro_rules! prim {
    ($fnname:ident $tname:ident) => {
        fn $fnname(&self) -> TypeRef<'a> {
            TypeRef(&Type::$tname)
        }
    };
}

impl<'a, W: WriteStr> TranslationUnitBuilder<'a, W> {
    prim!(bool Bool);
    prim!(u8 U8);
    prim!(i8 I8);
    prim!(u16 U16);
    prim!(i16 I16);
    prim!(u32 U32);
    prim!(i32 U32);
    prim!(u64 U64);
    prim!(i64 I64);
    prim!(float Float);
    prim!(double Double);
    prim!(intptr IntPtr);
    prim!(uintptr UintPtr);

    pub fn void(&self) -> IncompleteTypeRef<'a> {
        IncompleteTypeRef(&Type::Void)
    }

    /// Builds a pointer type pointing to `pointee`.
    pub fn ptr_to<Ty: AsType<'a>>(&mut self, pointee: Ty) -> TypeRef<'a> {
        TypeRef(self.arena.alloc(Type::Pointer(pointee.as_type())))
    }

    pub fn fn_sig(&mut self, ret: Option<TypeRef<'a>>, args: &[TypeRef<'a>]) -> FnSig<'a> {
        let ret = ret.unwrap_or(TypeRef(&Type::Void));
        FnSig {
            ret,
            args: self.arena.alloc_slice(args),
        }
    }

    pub fn fn_ptr(&mut self, sig: FnSig<'a>) -> TypeRef<'a> {
        TypeRef(self.arena.alloc(Type::FunctionPointer(sig)))
    }

    /// Emits a forward-declaration of a struct `inst`.
    pub fn fwd_declare_struct(&mut self, sym: Name<'_>) -> io::Result<IncompleteTypeRef<'a>> {
        writeln!(
            self.writer,
            "struct {};  /* {} */",
            sym.mangled(),
            sym.rust()
        )?;
        let name = self.arena.alloc_str(&sym.mangled());
        Ok(IncompleteTypeRef(
            self.arena.alloc(Type::FwdStruct { name }),
        ))
    }

    /// Emits a forward-declaration of a union `inst`.
    pub fn fwd_declare_union(&mut self, sym: Name<'_>) -> io::Result<IncompleteTypeRef<'a>> {
        writeln!(
            self.writer,
            "union {};  /* {} */",
            sym.mangled(),
            sym.rust()
        )?;
        let name = self.arena.alloc_str(&sym.mangled());
        Ok(IncompleteTypeRef(self.arena.alloc(Type::FwdUnion { name })))
    }

    pub fn fwd_declare_function(
        &mut self,
        name: Name<'_>,
        proto: FnSig<'a>,
    ) -> io::Result<Function<'a>> {
        proto.declare(&*name.mangled(), &mut self.writer)?;
        let fnty = self.fn_ptr(proto);
        writeln!(self.writer, "; /* {} */", name.rust())?;
        Ok(Function {
            sig: proto,
            ptr_ty: fnty,
            name: self.arena.alloc_str(&name.mangled()),
        })
    }

    fn define_named_type<'f, F>(
        &mut self,
        kind: &str,
        sym: Name<'_>,
        fields: F,
    ) -> io::Result<TypeRef<'a>>
    where
        F: IntoIterator<Item = (&'f str, TypeRef<'a>)>,
    {
        if self.defd_types.contains(&*sym.mangled()) {
            return Err(io::Error::new(
                io::ErrorKind::Other,
                format!("type '{}' already defined", sym.mangled()),
            ));
        }

        writeln!(
            self.writer,
            "{} {} {{  /* {} */",
            kind,
            sym.mangled(),
            sym.rust()
        )?;
        for (field, ty) in fields {
            write!(self.writer, "    ")?;
            ty.declare_variable(field, &mut self.writer)?;
            writeln!(self.writer, ";")?;
        }
        writeln!(self.writer, "}};")?;

        let name = self.arena.alloc_str(&sym.mangled());
        self.defd_types.insert(name);
        Ok(TypeRef(self.arena.alloc(Type::Struct { name })))
    }

    /// Defines a `struct` with field names and types.
    // TODO attach attributes
    pub fn define_struct<'f, F>(&mut self, sym: Name<'_>, fields: F) -> io::Result<TypeRef<'a>>
    where
        F: IntoIterator<Item = (&'f str, TypeRef<'a>)>,
    {
        self.define_named_type("struct", sym, fields)
    }

    /// Defines a `union` with field names and types.
    pub fn define_union<'f, F>(&mut self, sym: Name<'_>, fields: F) -> io::Result<TypeRef<'a>>
    where
        F: IntoIterator<Item = (&'f str, TypeRef<'a>)>,
    {
        self.define_named_type("union", sym, fields)
    }

    /// Starts defining a function, returning a `FunctionBuilder`.
    pub fn define_function<'b>(
        &'b mut self,
        arena: &'b Arena,
        name: Name<'_>,
        proto: FnSig<'a>,
    ) -> io::Result<FunctionBuilder<'b, W>> {
        Ok(FunctionBuilder::create(
            &mut self.writer,
            arena,
            name,
            proto,
        )?)
    }
}

/// A declared function.
pub struct Function<'a> {
    sig: FnSig<'a>,
    ptr_ty: TypeRef<'a>,
    name: &'a str,
}

#[cfg(test)]
mod tests {
    use super::test::compile_test;
    use super::*;
    use utils::StringWriter;

    use cc::Build;
    use std::error::Error;
    use tempfile::{self, TempDir};

    #[test]
    fn fwd_declare() {
        compile_test("fwd_declare", |f| {
            let i = f.i32();
            let _p = f.ptr_to(i);
            let v = f.void();
            let _p = f.ptr_to(v);

            f.fwd_declare_struct(Name::test("my_struct"))?;
            f.fwd_declare_union(Name::test("my_union"))?;
            Ok(())
        });
    }

    #[test]
    #[should_panic(expected = "already defined")]
    fn redefinition() {
        // A test that's supposed to fail has to panic inside the closure, or
        // else `test` will compile the output.
        compile_test("redefinition", |f| {
            let bool = f.bool();
            f.define_struct(Name::test("my_struct"), vec![("field1", bool)])?;
            let msg = f
                .define_struct(Name::test("my_struct"), vec![("field2", bool)])
                .unwrap_err()
                .to_string();
            panic!("{}", msg);
        });
    }

    #[test]
    fn define_struct() {
        compile_test("define_struct", |f| {
            let i = f.i32();
            let pi = f.ptr_to(i);
            let v = f.void();
            let pv = f.ptr_to(v);
            let b = f.bool();

            f.fwd_declare_struct(Name::test("my_struct"))?;
            f.fwd_declare_union(Name::test("my_union"))?;

            f.define_struct(
                Name::test("my_struct"),
                vec![
                    ("intfield", i),
                    ("intptr", pi),
                    ("voidptr", pv),
                    ("shevik", b),
                ],
            )?;
            Ok(())
        });
    }

    /// Defines a struct containing function pointer fields.
    #[test]
    fn define_struct_fn_ptr() {
        compile_test("define_struct_fn_ptr", |f| {
            let i = f.i32();
            let pi = f.ptr_to(i);
            let v = f.void();
            let pv = f.ptr_to(v);
            let b = f.bool();
            let fn_sig = f.fn_sig(Some(pi), &[i, i, b]);
            let fn_ptr = f.fn_ptr(fn_sig);
            let fn_sig = f.fn_sig(None, &[i, i, b]);
            let fn_ptr2 = f.fn_ptr(fn_sig);
            let fn_sig = f.fn_sig(Some(i), &[]);
            let fn_ptr3 = f.fn_ptr(fn_sig);

            f.define_struct(
                Name::test("my_struct"),
                vec![("f1", fn_ptr), ("f2", fn_ptr2), ("f3", fn_ptr3)],
            )?;
            Ok(())
        });
    }

    #[test]
    fn fwd_declare_fn() {
        compile_test("fwd_declare_fn", |f| {
            let i = f.i32();
            let pi = f.ptr_to(i);
            let b = f.bool();

            let sig = f.fn_sig(Some(pi), &[i, i, b]);
            f.fwd_declare_function(Name::test("myfn"), sig)?;
            let sig = f.fn_sig(None, &[]);
            f.fwd_declare_function(Name::test("empty"), sig)?;
            Ok(())
        });
    }

    #[test]
    fn def_fn() {
        compile_test("def_fn", |f| {
            let i = f.i32();
            let pi = f.ptr_to(i);
            let b = f.bool();

            // TODO: Test more way to declare fns (returning fn ptr, nothing, taking no args, ...)
            let sig = f.fn_sig(Some(pi), &[i, i, b]);
            let arena = Arena::new();
            let b = f.define_function(&arena, Name::test("myfn"), sig)?;
            b.finish()?;
            Ok(())
        });
    }
}
