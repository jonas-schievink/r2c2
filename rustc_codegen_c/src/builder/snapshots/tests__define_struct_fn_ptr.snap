---
created: "2019-02-07T00:21:01.110978750Z"
creator: insta@0.6.1
source: rustc_codegen_c/src/builder/mod.rs
expression: out
---
/* generated by r2c2 version 0.0.0 */
#include <stdint.h>
#include <stdbool.h>

/*                                                    */
/*                    END PREAMBLE                    */
/*                                                    */

struct my_struct {  /* (test-symbol: no associated Rust name) */
    uint32_t * (*f1)(uint32_t _1, uint32_t _2, bool _3);
    void (*f2)(uint32_t _1, uint32_t _2, bool _3);
    uint32_t (*f3)(void);
};

